src/navigation/water_bottle_search.py

```diff
@@ -0,0 +1,426 @@
+        :return: reversed path except the starting point (we are already there)
+        """
+        costmap2D = np.copy(costmap2D)
+        path = []
+        current = current_node
+        while current is not None:
+            path.append(current.position)
+            current = current.parent
+        reversed_path = path[::-1]
+        print("ij:", reversed_path[1:])
+
+        # Print visual of costmap with path and start (S) and end (E) points
+        # The lighter the block, the more costly it is
+        for step in reversed_path:
+            costmap2D[step[0]][step[1]] = 2 # path (.)
+        costmap2D[reversed_path[0][0]][reversed_path[0][1]] = 3  # start
+        costmap2D[reversed_path[-1][0]][reversed_path[-1][1]] = 4  # end
+
+        for row in costmap2D:
+            line = []
+            for col in row:
+                if col == 1.0:
+                    line.append("\u2588")
+                elif col < 1.0 and col >= 0.8:
+                    line.append("\u2593")
+                elif col < 0.8 and col >= 0.5:
+                    line.append("\u2592")
+                elif col < 0.5 and col >= 0.2:
+                    line.append("\u2591")
+                elif col < 0.2:
+                    line.append(" ")
+                elif col == 2:
+                    line.append(".")
+                elif col == 3:
+                    line.append("S")
+                elif col == 4:
+                    line.append("E")
+            print("".join(line))
+
+        return reversed_path[1:]
+
+    def a_star(self, costmap2d: np.ndarray, start: np.ndarray, end: np.ndarray) -> list | None:
+        """
+        A-STAR Algorithm: f(n) = g(n) + h(n) to find a path from the given start to the given end in the given costmap
+        :param costmap2d: occupancy grid in the form of an 2D array of floats
+        :param start: rover pose in cartesian coordinates
+        :param end: next point in the spiral from traj in cartesian coordinates
+        :return: list of A-STAR coordinates in the occupancy grid coordinates (i,j)
+        """
+        # convert start and end to occupancy grid coordinates
+        startij = self.cartesian_to_ij(start)
+        endij = self.cartesian_to_ij(end)
+
+        # initialize start and end nodes
+        start_node = Node(None, (startij[0], startij[1]))
+        end_node = Node(None, (endij[0], endij[1]))
+
+        if start_node == end_node:
+            return None
+
+        print(f"start: {start}, end: {end}")
+        print(f"startij: {startij}, endij: {endij}")
+
+        # check if end node is within range, if it is, check if it has a high cost
+        if (
+            end_node.position[0] <= (costmap2d.shape[0] - 1)
+            and end_node.position[0] >= 0
+            and end_node.position[1] <= (costmap2d.shape[1] - 1)
+            and end_node.position[1] >= 0
+        ):
+            while(costmap2d[end_node.position[0], end_node.position[1]] >= 0.2): # TODO: find optimal value
```

Comment:
move this to a function that computes the end point  separately (in the State class) and then pass that endpoint into the a_star function. the a_star function shouldn't change the end point